### DISCUSSION TITLE
CommandsNext refactor

### DISCUSSION THREAD

**Discussion by Kaoticz** [DSharpPlus/DSharpPlus#1245] [General]
I've been wanting to add some features to [DSharpPlus.CommandsNext](https://github.com/DSharpPlus/DSharpPlus/tree/master/DSharpPlus.CommandsNext) command framework, but after taking a quick glance at the code I realized that implementing new features would be near impossible due to the insane amount of code debt present in it.

I don't object the idea of doing a major refactor just to get the project to a "workable" state, but before I sink a significant amount of time into it, I'd like to know if you guys would accept such changes. I'm aware there is work currently being done on the next version of D#+ (V5), which would probably warrant a complete rewrite of the command framework anyway, but that version is a long way off (and I kinda need these features now 😅). The refactor would be implemented in the following phases (input on these is appreciated):

- Enable [NRT](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references).
- Improve readability (get rid of crappy variable names, replace verbose patterns with simpler ones).
- Code cleanup (remove unneeded variables, unneeded method calls, etc).
- Code re-architecture (potential breaking change - rename internal methods/types, move stuff around, create new types and abstractions wherever applicable, etc).
- Investigate the feasibility of doing the same to [DSharpPlus.SlashCommands](https://github.com/DSharpPlus/DSharpPlus/tree/master/DSharpPlus.SlashCommands) and merging it into [DSharpPlus.CommandsNext](https://github.com/DSharpPlus/DSharpPlus/tree/master/DSharpPlus.CommandsNext).
- Create unit tests.

After the work is done, I hope that working on the command framework should become significantly easier (which would also make porting it to V5 easier, whenever it comes out).

What are your thoughts on this?

**Comment by akiraveliara** [comment]
okay, sooo... generally i agree, personally i don't have the motivation to work on commands for v4.

itd probably get rejected if you made one huge PR, but several smaller PRs should be fine, from what i've seen so far.

with that said;

- NRTs are unfortunately not happening (afaik) because v4 strictly targets .NET Standard,
- with the slash commands extension largely unmaintained, you'd be pretty much entirely on your own there,
- v5 cnext shouldnt be a port of v4 cnext in the first place, instead, a proper rewrite

**Reply by OoLunar** [reply]
> NRTs are unfortunately not happening (afaik) because v4 strictly targets .NET Standard,

By this, I think they mean enabling `Nullable` in the csproj, in which case can definitely be done since I'm currently doing it in the VNext "rewrite" (which may never get done).

> v5 cnext shouldnt be a port of v4 cnext in the first place, instead, a proper rewrite

Everything in v5 is a rewrite, there shouldn't be ported code unless if it's quality code.

**Reply by Kaoticz** [reply]
Yeah, that's what I had in mind as well. Make very targeted changes, create a PR, let it get approved, then work on the next improvements. I'm also not adding or removing any features, which should make the review process easier.

> NRTs are unfortunately not happening (afaik) because v4 strictly targets .NET Standard

Lunar is right, I've enabled nullable annotations on CommandsNext without changing the target project and I confirm that it works as intended.

> with the slash commands extension largely unmaintained, you'd be pretty much entirely on your own there

That's fine, just refactoring it won't be much of an issue, and most of the work I'd do on it would occur during the integration with CommandsNext (which I don't know if it's going to happen).

> v5 cnext shouldnt be a port of v4 cnext in the first place, instead, a proper rewrite

Alright, that's fair. Considering how messy the current code is, I can see why everyone would rather write a command framework from scratch. But if I do that, my PRs won't get accepted 😄

**Reply by OoLunar** [reply]
For what it's worth, I spent most of today writing a CONTRIBUTING.md file and a "Development Plan.md" on the v5 branch. They're uncommitted, but I'll let you know when they're there.

**Comment by OoLunar** [comment]
Exal you sniper.

--

Well, everything you've mentioned is exactly what we want to do with v5. Here's how we plan to make v5:
Create DSharpPlus.Core:
- Extract the Discord objects into C# classes. These make use of `Optional` structs, such as `Optional<int?>` and will not have a cache
- Implement the Discord Client
- Create DSharpPlus.Abstractions. This involves convince/computed properties such as `DefaultAvatarUrl` and `BannerUrl`, and will most certainly have a cache. `Optional<int?>` will turn into `int?`
- Rewrite extensions
Now that we have an actual roadmap ahead of us which contains a decent way to satisfy everyone, implementing things should be extremely easy. If anything, the next stopping point that v5 will get stuck at is extensions, in which we've had many people verbally volunteer to help with. In the mean time, if you feel you can rewrite CNext in a timely manner (that is, within the next 6 months) and your PR is a good quality PR, there is a small chance that we would accept a CNext rewrite with breaking API changes, and a large chance of accepting a CNext internal cleanup with feature additions. If you feel like you can rewrite the extension twice, go you, however I feel your work would be more appreciated with the v5 rewrite. Obviously beggars can't be choosers, so do as you like.

> Investigate the feasibility of doing the same to [DSharpPlus.SlashCommands](https://github.com/DSharpPlus/DSharpPlus/tree/master/DSharpPlus.SlashCommands) and merging it into [DSharpPlus.CommandsNext](https://github.com/DSharpPlus/DSharpPlus/tree/master/DSharpPlus.CommandsNext).

Absolutely. I want to have 3 separate packages: CNext, SlashCommands, and MixedCommands. MixedCommands would find common ground between CNext and Slashies, with special attributes to make specific usage of either extensions. I haven't actually discussed this idea, however this isn't the first time the idea has been brought up.

**Reply by akiraveliara** [reply]
had to bully me by adding that l, i see how it is :( /lh

**Reply by akiraveliara** [reply]
re: command packages, those should inherit a generic attribute & interface package so we dont end up with `DSharpPlus.CNext.Attributes.SomeAttribute` and `DSharpPlus.Slashies.Attributes.SomeAttribute` again

**Reply by OoLunar** [reply]
That's a good point. I figured I was going to explore this more when it's time to implement it.

**Reply by Kaoticz** [reply]
Alright, so I gave this some thought and one of your points caught my attention. You're suggesting that V5 will near completion after 6 months from now. Since development has started just a month ago, I assumed the final version would only be ready much later (over a year at least). If V5 is getting completed in this time frame, then I agree that working on the current version of CNext is going to be a waste of time and effort, even if the refactor is completed in a timely manner.

I could live with some workarounds if you believe V5 can be completed in less than a year. However, I'd still like to at least enable nullable annotations on both CNext and SlashCommands, so I can get feedback on nullability directly from my projects (I've had my share of unpleasant surprises with sneaky NREs).

For now, I think I'm going to do that, then reassess the situation and my next actions.
I could help with V5 with some little bits here and there. Compiling a list of things to do like [this](https://github.com/DSharpPlus/DSharpPlus/issues/825) and defining some guidelines for the general design of the API would be of great help for those who are interested in contributing.

**Reply by OoLunar** [reply]
> You're suggesting that V5 will near completion after 6 months from now.

It's only a goal with zero promises. Rumors of v5 span back for years now, so I'm sure there are other people who may bring up ideas or concepts which completely go against the current development plan, thus pushing back development. Of course, I'm going to try to avoid that.

> However, I'd still like to at least enable nullable annotations on both CNext and SlashCommands, so I can get feedback on nullability directly from my projects (I've had my share of unpleasant surprises with sneaky NREs).

If you create a PR and fix the compiler warnings, I'll go through and test it, likely merging it.

> I could help with V5 with some little bits here and there. Compiling a list of things to do like https://github.com/DSharpPlus/DSharpPlus/issues/825 and defining some guidelines for the general design of the API would be of great help for those who are interested in contributing.

Spent all day yesterday writing a `CONTRIBUTING.md`, going to spend some time today finishing the final details of `Development Plan.md`
